contracts/CharityChallenge.sol

```diff
@@ -89,8 +103,19 @@ contract CharityChallenge {
             isEventFinalizedAndValid = true;
             if (hasChallengeAccomplished) {
                 uint256 totalContractBalance = address(this).balance;
-                npoAddress.transfer(address(this).balance);
-                emit Donated(npoAddress, totalContractBalance);
+                uint8 sumRatio = 0;
+                uint length = npoAddresses.length;
+                for (uint i = 0; i < length; i++) {
+                  address payable npo = npoAddresses[i];
+                  sumRatio += npoRatios[npo];
+                }
+                for (uint i = 0; i < length; i++) {
+                  address payable npo = npoAddresses[i];
+                  uint8 ratio = npoRatios[npo];
+                  uint256 amount = totalContractBalance * ratio / sumRatio;
+                  npo.transfer(amount);
```

Comment:
can we determine the payout amount in the constructor and then set
`mapping (address => payoutAmount)`

Reply:
pros: we can verify how much each address would have right after we deploy

Reply:
nvm, it should not in the constructor because the balance is 0. The goal is to verify the expected payout as soon as the contract has balance.

Reply:
I added getExpectedDonationAmount